### PR TITLE
Add `HashLiteral#has_key?` and `NamedTupleLiteral#has_key?`

### PR DESCRIPTION
**Pull request by kamil-gwozdz**
This PR adds `HashLiteral#has_key?` and `NamedTupleLiteral#has_key?`.

closes https://github.com/crystal-lang/crystal/issues/14841

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
Thanks for the PR! In the future, as called out in https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#making-good-pull-requests, could you not force push? Doesn't matter too much in this case as it was a WIP branch, but it can make the review process harder if force pushes happen with pending comments and such. We squash merge PRs so no need to worry about having a lot of commits.

File: src/compiler/crystal/macros/methods.cr
Comment (Sija) [inline src/compiler/crystal/macros/methods.cr:1095]: Please, format the exception message similarly to the other instances.

Comment (kamil-gwozdz) [reply src/compiler/crystal/macros/methods.cr:1095]: i copied it from here: https://github.com/crystal-lang/crystal/pull/14890/files#diff-5d087db61d41eddf80efd1bb7dae3dc22c81b6ebe9e98cc36606480519248b6bR1057. I didn't notice that some other exceptions are using the `klass#method` notation.

BTW it seems that out of 39[^1] exceptions 6[^2] aren't referencing the class. Is this something that should be addressed in a followup PR? 


[^1]: `git grep "raise.*argument to " | wc -l`
[^2]: 39 - `git grep "raise.*argument to " | grep "#" | wc -l` = 6

Comment (kamil-gwozdz) [reply src/compiler/crystal/macros/methods.cr]: i reformatted the exception message: 221aaedf316b7c5248886c5342ead6323959590d

File: src/compiler/crystal/macros/methods.cr
Comment (Sija) [inline src/compiler/crystal/macros/methods.cr:1095]: My suggestion:

```suggestion
            raise "expected 'NamedTupleLiteral#has_key?' first argument to be a SymbolLiteral or StringLiteral, not #{key.class_desc}"

```

Comment (kamil-gwozdz) [reply src/compiler/crystal/macros/methods.cr]: i also added the `MacroId` to the list since it's in the case above

File: src/compiler/crystal/macros/methods.cr
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr:971]: **suggestion:** `entry.nil?` is more idiomatic than `entry == nil`. `#==` could theoretically be overridden, `#nil?` cannot.
```suggestion
          BoolLiteral.new(!entry.nil?)
```

File: src/compiler/crystal/macros/methods.cr
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr:1099]: Ditto.
```suggestion
          BoolLiteral.new(!entry.nil?)
```

File: src/compiler/crystal/macros/methods.cr
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr:1093]: **polish:** I think these branches could be merged into one for a bit more concise expression:
```suggestion
          when SymbolLiteral, MacroId, StringLiteral
            key = key.value
```

Comment (kamil-gwozdz) [reply src/compiler/crystal/macros/methods.cr]: 👍🏻 

i also changed it in the places where i copied it from: 1f30fc6d91c944669bc79f158b236b5384f4c3be

**Review by Sija — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

**Comment by kamil-gwozdz** [comment]
Thanks for the reviews. Unfortunately, i can't push to `crystal-lang:master`. Can somebody merge this PR for me?

**Comment by straight-shoota** [comment]
We'll merge the PR when it gets accepted. For that we still need an approval from another Core Team member.
See https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#accepting-a-pull-request for the workflow.

File: src/compiler/crystal/macros/methods.cr
Comment (HertzDevil) [inline src/compiler/crystal/macros/methods.cr:971]:
```suggestion
          BoolLiteral.new(entries.any? &.key.==(key))
```

File: src/compiler/crystal/macros/methods.cr
Comment (HertzDevil) [inline src/compiler/crystal/macros/methods.cr:1087]:
```suggestion
          BoolLiteral.new(entries.any? &.key.==(key))
```

**Review by ysbaddaden — APPROVED** [review]

**Pull request merged**
